### PR TITLE
Add support for MeterEntry in Stratum-bfrt

### PR DESCRIPTION
**Pull request by pudelkoM**
This PR adds support for P4RT `MeterEntry` in Stratum-bfrt.

Snippet:

```python
te = table_entry['FabricIngress.spgw.qos_classifier'](action='FabricIngress.spgw.set_qid_midx')
te.match['ue_addr'] = "10.0.0.1"
te.action['qid'] = '1'
te.action['meter_idx'] = '5'
te.priority = 10
te.insert()
te.read(lambda e: print(e))
me = meter_entry['FabricIngress.spgw.meter']
me.index = 5
me.config.cir = 1
me.config.pir = 2
me.config.cburst = 100
me.config.pburst = 200
me.modify()
me.read(lambda e: print(e))
```

We observed some potential issues around rw symmetry around  `MeterConfig`. The SDE will report back the actually programed values from the ASIC. Due to optimizations/limitations these can be slightly different from the requested values:

```python
me = meter_entry['FabricIngress.spgw.meter']
me.index = 5
me.config.cir = 8000000
me.config.pir = 8000000
me.config.cburst = 800000
me.config.pburst = 800000
me.modify()
me.read(lambda e: print(e))
meter_id: 335565318 ("FabricIngress.spgw.meter")
index {
  index: 5
}
config {
  cir: 7992000
  cburst: 800000
  pir: 7992000
  pburst: 800000
}
```

### CODE REVIEW

File: stratum/hal/lib/barefoot/bf_sde_wrapper.h [hunk 1]
```diff
@@ -99,6 +99,7 @@ class BfSdeWrapper : public BfSdeInterface {
  public:
   // Default MTU for ports on Tofino.
   static constexpr int32 kBfDefaultMtu = 10 * 1024;  // 10K
+  static constexpr char kFoo[] = "foo";
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_sde_wrapper.h:102]: remove

File: stratum/hal/lib/barefoot/bf_sde_wrapper.cc [hunk 2]
```diff
@@ -2208,6 +2208,172 @@ ::util::Status BfSdeWrapper::ReadRegisters(
...
+    // Data: $METER_SPEC_*
+    std::vector<bf_rt_id_t> data_field_ids;
+    RETURN_IF_BFRT_ERROR(table->dataFieldIdListGet(&data_field_ids));
+    for (const auto& field_id : data_field_ids) {
+      std::string field_name;
+      RETURN_IF_BFRT_ERROR(table->dataFieldNameGet(field_id, &field_name));
+      if (field_name == kMeterCirKbps) {  // In kbits
+        uint64 cir;
+        RETURN_IF_BFRT_ERROR(table_data->getValue(field_id, &cir));
+        cirs->push_back(cir * 125);
```
Comment (Yi-Tseng) [inline stratum/hal/lib/barefoot/bf_sde_wrapper.cc:2328]: add comments like `Convert SDE kbits to P4RT bytes (* 1000 / 8).`

File: stratum/hal/lib/barefoot/bfrt_table_manager.cc [hunk 3]
```diff
@@ -829,6 +831,118 @@ ::util::Status BfrtTableManager::WriteRegisterEntry(
...
+      device_, session, table_id, optional_meter_index, &meter_indices, &cirs,
+      &cbursts, &pirs, &pbursts, &in_pps));
+
+  ::p4::v1::ReadResponse resp;
+  for (size_t i = 0; i < meter_indices.size(); ++i) {
+    const uint32 meter_index = meter_indices[i];
+    const uint64 cir = cirs[i];
+    const uint64 cburst = cbursts[i];
+    const uint64 pir = pirs[i];
+    const uint64 pburst = pbursts[i];
```
Comment (Yi-Tseng) [inline stratum/hal/lib/barefoot/bfrt_table_manager.cc:881]: is it necessary to have variables here?

**Review by Yi-Tseng — APPROVED** [review]

File: stratum/hal/lib/barefoot/bfrt_table_manager.cc (hunk 3 above)
Comment (pudelkoM) [reply stratum/hal/lib/barefoot/bfrt_table_manager.cc:881]: not really, it's a copy&paste leftover

File: stratum/hal/lib/barefoot/bfrt_table_manager.cc [hunk 4]
```diff
@@ -829,6 +831,118 @@ ::util::Status BfrtTableManager::WriteRegisterEntry(
...
+                     p4_info_manager_->FindMeterByID(meter_entry.meter_id()));
+    switch (meter.spec().unit()) {
+      case ::p4::config::v1::MeterSpec::BYTES:
+        meter_units_in_packets = false;
+        break;
+      case ::p4::config::v1::MeterSpec::PACKETS:
+        meter_units_in_packets = true;
+        break;
+      default:
+        RETURN_ERROR(ERR_INVALID_PARAM) << "Unsupported meter spec on meter"
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bfrt_table_manager.cc:926]:
```suggestion
        RETURN_ERROR(ERR_INVALID_PARAM) << "Unsupported meter spec on meter "
```

File: stratum/hal/lib/barefoot/bf_sde_wrapper.cc (hunk 2 above)
Comment (pudelkoM) [reply stratum/hal/lib/barefoot/bf_sde_wrapper.cc]: Fixed by introducing helper functions.

File: stratum/hal/lib/barefoot/bf_sde_wrapper.cc [hunk 5]
```diff
@@ -293,11 +299,20 @@ ::util::Status GetAllEntries(
   CHECK_RETURN_IF_FALSE(table_keys) << "table_keys is null";
   CHECK_RETURN_IF_FALSE(table_datums) << "table_datums is null";
 
-  // Get number of entries.
+  // Get number of entries. Some types of tables are preallocated and are always
+  // "full". The SDE does not support querying the useage for these.
```
Comment (pudelkoM) [inline stratum/hal/lib/barefoot/bf_sde_wrapper.cc:303]:
```suggestion
  // "full". The SDE does not support querying the usage on these.
```

**Pull request merged**